vimode/src/cmds/motion.c

```diff
@@ -40,42 +40,53 @@ void cmd_goto_right(CmdContext *c, CmdParams *p)
 	SET_POS(p->sci, pos, TRUE);
 }
 
+static gint doc_line_from_visible_delta(CmdParams *p, gint line, gint delta, gint *previous)
+{
+	gint step = delta < 0 ? -1 : 1;
+	gint new_line = p->line;
+	gint left = abs(delta);
+	gint prev_line = -1, tmp;
+
+	if ((step > 0 && new_line < p->line_num - 1) || (step < 0 && new_line > 0)) {
+		while (left > 0) {
+			tmp = new_line;
+			if (SSM(p->sci, SCI_GETLINEVISIBLE, new_line, 0)) {
+				left--;
+				prev_line = tmp;
+			}
+			new_line += step;
+			if (new_line >= p->line_num || new_line <= 0) break;
+		}
+	}
+
+	if (previous) *previous = prev_line;
+
+	return new_line;
+}
 
 void cmd_goto_up(CmdContext *c, CmdParams *p)
 {
-	gint one_above, pos;
+	gint previous = -1;
+	gint new_line;
+	gint wrap_count;
 
 	if (p->line == 0)
 		return;
 
+	new_line = doc_line_from_visible_delta(p, p->line, -p->num, &previous);
+
 	/* Calling SCI_LINEUP/SCI_LINEDOWN in a loop for num lines leads to visible
 	 * slow scrolling. On the other hand, SCI_LINEUP preserves the value of
 	 * SCI_CHOOSECARETX which we cannot read directly from Scintilla and which
 	 * we want to keep - perform jump to previous/following line and add
 	 * one final SCI_LINEUP/SCI_LINEDOWN which recovers SCI_CHOOSECARETX for us. */
-	one_above = p->line - p->num - 1;
-	if (one_above >= 0 && SSM(p->sci, SCI_GETLINEVISIBLE, one_above, 0))
-	{
-		/* Every case except for the first line - go one line above and perform
-		 * SCI_LINEDOWN. This ensures that even with wrapping on, we get the
-		 * caret on the first line of the wrapped line */
-		pos = SSM(p->sci, SCI_GETLINEENDPOSITION, one_above, 0);
+
+	if (previous > -1) {
+		guint pos = SSM(p->sci, SCI_POSITIONFROMLINE, previous, 0); 
```

Comment:
I think you didn't understand the purpose of this strange "one above" and then `SCI_LINEDOWN` dance. While you get to the correct line alright, you lose the cursor's `x` position on the line this way. Notice how, when moving cursor up and down, it remembers the maximum `x` coordinate on the line and even after bing on lines where the maximum `x` is lower *like e.g. an empty line where `x` is 0), it returns back to the right position on longer lines.

Unfortunately this "maximum x" is not possible to obtain from Scintilla but Scintilla automatically recovers it when doing `SCI_LINEDOWN` or `SCI_LINEUP`. So the trick here is to first go to the line above or below, and then perform `SCI_LINEDOWN` or `SCI_LINEUP` to get us to the right line and get the `x` position of the cursor. When you remove this code, you'll always end with the `x` position at the beginning of the line.

Reply:
Yes, I think I understood the problem you explained correctly.

But from what I understood, for cmd_goto_up this method is only really necessary when the line you want to go to is not the previous visible line.

With this fix my idea was to have "doc_line_from_visible_delta" fill the "previous" variable with the line number just below the visible line to access it with a SET_POS_NOX, then in any case do a SCI_LINEUP to go to the desired line.

Same for goto_down and SCI_LINEDOWN.

Reply:
OK, so it was me who misunderstood your code :-).

Even though your version works, I'll probably go for https://github.com/geany/geany-plugins/pull/1338 which is easier for me to understand - unless you have some problems with it.